File: GUILib/src/space/kiibou/net/common/Callbacks.java
package space.kiibou.net.common;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Callbacks<IN, OUT> {

    private static long handleCounter = 0;

    private final Map<Long, Function<IN, OUT>> callbacks;

    public Callbacks() {
        callbacks = Collections.synchronizedMap(new HashMap<>());
    }

    private static long nextHandle() {
        return handleCounter++;
    }

    public long addCallback(final Function<IN, OUT> function) {
        Objects.requireNonNull(function);

        long handle = nextHandle();
        callbacks.put(handle, function);
        return handle;
    }

    public Function<IN, OUT> removeCallback(final long handle) {
        return callbacks.remove(handle);
    }

    public OUT call(final long handle, final IN arg) {
        return callbacks.get(handle).apply(arg);
    }

    public List<OUT> callAll(final IN arg) {
        return callbacks.values().stream()
                .map(fun -> fun.apply(arg))
                .collect(Collectors.toList());
    }

}

File: GUILib/src/space/kiibou/net/common/Callbacks.kt
package space.kiibou.net.common

import java.util.*
import kotlin.streams.toList

class Callbacks<R, S> {
    private val callbacks: MutableMap<Long, (R) -> S> = Collections.synchronizedMap(HashMap())

    fun addCallback(function: (R) -> S): Long {
        Objects.requireNonNull(function)
        val handle = nextHandle()
        callbacks[handle] = function
        return handle
    }

    fun removeCallback(handle: Long): ((R) -> S)? {
        return callbacks.remove(handle)
    }

    fun call(handle: Long, arg: R): S {
        return callbacks[handle]!!(arg)
    }

    fun callAll(arg: R): List<S> {
        return callbacks.values.stream()
                .map { `fun`: (R) -> S -> `fun`(arg) }
                .toList()
    }

    companion object {
        private var handleCounter: Long = 0
        private fun nextHandle(): Long {
            return handleCounter++
        }
    }

}